flight/mcu/lib/FifoBuffer/FifoBuffer.h

```diff
@@ -1,67 +1,123 @@
-class GenericFifoBuffer {
-   public:
-    // buff_size = size of array in elements
-    // data_size = size of array element in bytes
-    GenericFifoBuffer(void* arr, uint16_t buff_size, uint16_t data_size) {
-        capacity_ = buff_size;
-        cur_length_ = 0;
-        head_idx_ = 0;
-        tail_idx_ = 0;
-        data_size_ = data_size;
-        arr_ = arr;
-    }
+        FifoIterator(T* buf, size_t idx) : buf(buf), idx(idx) {}
 
-    bool push(void* element);
+        T& operator*() const {
+            return buf[idx];
+        }
+        T* operator->() {
+            return &buf[idx];
+        }
+        FifoIterator& operator++() {
+            idx = (idx + 1) % max_size;
+            return *this;
+        }
+        FifoIterator operator++(int) {
+            FifoIterator tmp = *this;
+            ++(*this);
+            return tmp;
+        }
+        friend bool operator== (const FifoIterator& a, const FifoIterator& b) {
+            return a.buf == b.buf && a.idx == b.idx;
+        }
+        friend bool operator!= (const FifoIterator& a, const FifoIterator& b) {
+            return a.buf != b.buf || a.idx != b.idx;
+        }
 
-    // returns false on failure
-    bool pop(void* out);
+    private:
+        T* buf;
+        size_t idx;
+    };
 
-   private:
-    uint16_t capacity_, cur_length_, head_idx_, tail_idx_, data_size_;
-    void* arr_;
 
-#ifdef COMPILE_LOCAL
-    std::mutex lock_;
-#endif
+    FifoBuffer() : length(0), head_idx(0), tail_idx(0) {}
 
-#ifdef COMPILE_TARGET
-    mutex_t lock_;
-#endif
-};
+    bool push(T& element) {
+        lock();
+        if (length == max_size) {
+            // if length == max_size, then head_idx == tail_idx
+            arr[tail_idx] = element;
+            head_idx = tail_idx = (head_idx + 1) % max_size;
+        } else {
+            arr[tail_idx] = element;
+            tail_idx = (tail_idx + 1) % max_size;
+            length++;
+        }
+        unlock();
+        return true;
+    }
 
-template <typename T, size_t max_size>
-class FifoBuffer {
-    // static_assert(std::is_trivially_copyable<T>::value, "Only trivially
-    // copyable types are allowed");
-   public:
-    FifoBuffer() : buffer(arr, max_size, sizeof(T)) {}
+    bool pop(T* out) {
+        lock();
+        if (length == 0) {
+            unlock();
+            return false;
+        } else {
+            *out = arr[head_idx];
+            head_idx = (head_idx + 1) % max_size;
+            length--;
+            unlock();
+            return true;
+        }
+    }
 
-    bool push(T element) { return buffer.push(&element); }
+    void lock() {
+        chMtxLock(&lock_);
+    }
+
+    void unlock() {
+        chMtxUnlock(&lock_);
+    }
+
+    FifoIterator begin() {
+        return FifoIterator { arr, head_idx };
+    }
 
-    bool pop(T* out) { return buffer.pop(out); }
+    FifoIterator end() {
+        return FifoIterator { arr, tail_idx };
+    }
+
+    std::vector<T> recentN(size_t n) {
```

Comment:
We can't use vector in flight code because it does dynamic allocation. This function should take a pointer to an array that it writes out to.

Reply:
Fixed, my bad